### PR TITLE
Added a V2 Reports Endpoint

### LINKED ISSUE #2679: Nest the report detail view inside the v2 asset endpoint
`/reports/` was probably added when it was difficult to write nested endpoints. Times have changed!

The `/reports/` list view seems never to be called by our front-end code, and it's basically an incomplete re-implementation of the asset list view that also filters for deployed assets and `view_submission`/`partial_submission` privileges. The front end always directly queries the `/reports/{asset uid}/` detail view.

Also, this is problematic **(and is now fixed by #2680)**:
https://github.com/kobotoolbox/kpi/blob/1acf9a54ae80816d5b2c5f087fdde45d4e3118d1/kobo/apps/reports/views.py#L35-L37
We shouldn't be serving up a list of all assets with public data, only the assets that have public data _and_ are inside a discoverable collection. Again, this is an incomplete redo of the asset list view: contrast with the handling of `all_public` in [`KpiObjectPermissionsFilter`](https://github.com/kobotoolbox/kpi/blob/8b930f5d871feafd31077327b311396d7108c2c4/kpi/filters.py#L31).

### PR DESCRIPTION
**Pull request by JacquelineMorrissette**
## Checklist

1. [x] If you've added code that should be tested, add tests
2. [x] If you've changed APIs, update (or create!) the documentation
3. [x] Ensure the tests pass
4. [x] Make sure that your code lints and that you've followed [our coding style](https://github.com/kobotoolbox/kpi/blob/master/CONTRIBUTING.md)
5. [x] Write a description of your work suitable for publishing on [our forum](https://community.kobotoolbox.org/tag/release-notes)
6. [x] Mention any related issues in this repository (as #ISSUE) and in other repositories (as kobotoolbox/other#ISSUE)

## Description

-An API endpoint for Reports v2 has been created.
-The new endpoint can be accessed via `https://[kpi]/api/v2/assets/{asset_uid}/reports/`

## Related issues

Fixes #2679

### CODE REVIEW

File: kpi/tests/api/v2/test_api_assets.py [hunk 1]
```diff
@@ -439,6 +439,61 @@ def test_report_styles_field(self):
...
+        response = self.client.get(report_url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # Verify the endoiubt returns data
+        self.assertEqual(response.data['count'], 1)
+        # Verify the endpoint request fails when the user is logged out
+        self.client.logout()
+        response = self.client.get(report_url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        # Verify a user with partial view permissions can access the data
+        self.client.login(username='anotheruser', password='anotheruser')
```
Comment (noliveleger) [inline kpi/tests/api/v2/test_api_assets.py:487]: You can use the same user with different permissions. 
We should test with `'view_submissions'` and `'partial_submissions'`

File: kpi/permissions.py [hunk 2]
```diff
@@ -273,3 +274,18 @@ class SubmissionValidationStatusPermission(SubmissionPermission):
...
 
 
+class ReportPermission(IsOwnerOrReadOnly):
+    def has_object_permission(self, request, view, obj):
+        # Checks if the user has the require permissions
+        # To access the submission data in reports
+        user = request.user
+        if user.is_anonymous:
+            user = get_anonymous_user()
+        permissions = list(obj.get_perms(user))
```
Comment (noliveleger) [inline kpi/permissions.py:284]: Let's add a validation if user is a super user and return `True`

```python
        if user.is_superuser:
            return True
            
        if user.is_anonymous:
            user = get_anonymous_user()
            
        permissions = list(obj.get_perms(user))
        
```

**Review by noliveleger — CHANGES_REQUESTED** [review]

File: kpi/tests/api/v2/test_api_assets.py [hunk 3]
```diff
@@ -439,6 +439,66 @@ def test_report_styles_field(self):
...
+        partial_perms = {
+            PERM_VIEW_SUBMISSIONS: [{'_submitted_by': anotheruser.username}]
+        }
+        self.asset.assign_perm(
+            partialaccess, PERM_PARTIAL_SUBMISSIONS, partial_perms=partial_perms
+        )
+        # Verify endpoint works with the asset owner
+        response = self.client.get(report_url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # Verify the endoiubt returns data
```
Comment (noliveleger) [inline kpi/tests/api/v2/test_api_assets.py:481]: Typo :-) I guess you meant `endpoint`

**Comment by JacquelineMorrissette** [comment]
I've made all your suggested changes @noliveleger and I agree that we should wait until after Beta is released before deploying the endpoint. 

I'm still having issues with the user that I created. I wasn't able to figure out why I'm getting 200_OK when the users that I created but when I test it manually on the API I get the 403. 

If you have any other changes or suggestions, please let me know!

File: kpi/tests/api/v2/test_api_assets.py (hunk 3 above)
Comment (JacquelineMorrissette) [reply kpi/tests/api/v2/test_api_assets.py]: Fixed!

File: kpi/permissions.py (hunk 2 above)
Comment (JacquelineMorrissette) [reply kpi/permissions.py:286]: Added the superuser to the permissions

File: kpi/tests/api/v2/test_api_assets.py (hunk 1 above)
Comment (JacquelineMorrissette) [reply kpi/tests/api/v2/test_api_assets.py:485]: Fixed!

File: kpi/tests/api/v2/test_api_assets.py [hunk 4]
```diff
@@ -439,6 +439,61 @@ def test_report_styles_field(self):
...
+        response = self.client.get(report_url)
+        self.assertEqual(response.status_code, status.HTTP_404_NOT_FOUND)
+        # Verify a user with partial view permissions can access the data
+        self.client.login(username='anotheruser', password='anotheruser')
+        response = self.client.get(report_url)
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        # Verify a user with no permissions for the submissions or asset cannot
+        # access the data
+        self.client.logout()
+        self.client.login(username='admin', password='pass')
```
[the remark this answers is not in the document]
Comment (JacquelineMorrissette) [reply kpi/tests/api/v2/test_api_assets.py:501]: The admin will now be able to view all users submission data

**Comment by noliveleger** [comment]
> I'm still having issues with the user that I created. I wasn't able to figure out why I'm getting 200_OK when the users that I created but when I test it manually on the API I get the 403.

Because of that:
https://github.com/kobotoolbox/kpi/pull/3009/commits/b95237da4da87bc7fcc3922619b2aa97954b7827#diff-623e2feaabba7cf370ce46f406624efdf99c6a47b7d6bff5bd3a79ed69b5d921L475-L477

I did the test
```python
        partialaccess = User.objects.create(username='partialaccess',
                                            password='partialaccess')
        self.client.login(username='partialaccess', password='partialaccess')
        response = self.client.get(report_url)
        self.assertEqual(response.status_code, status.HTTP_200_OK)
``` 
It did fail

**Pull request merged**